### ISSUE TITLE
CA2227 false positive on record type

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn-analyzers#4461]
### Analyzer

**Diagnostic ID**: [CA2227](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2227): `Collection properties should be read only`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

The analyzer incorrectly reports a CA2227 for a list property of a record.

### Steps To Reproduce

Write this code in a project where CA2227 is enabled

```csharp
public record MyRecord(IList<Item> Items);
```

### Expected behavior

No warning

### Actual behavior

CA2227 warning on `Items`:

> Change 'Items' to be read-only by removing the property setter

But here the setter is an `init` accessor (compiler-generated), which means the property can't be reassigned after creation.

## Additional context

The problem also occurs for properties with init-only setters on non-record types.
The analyzer should look for a `modreq` of type `IsExternalInit` on the setter's return parameter, as specified in the [init-only setter proposal](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/init)

**Comment by Evangelink** [comment]
I confirm the issue. 

Currently the version of Roslyn we target doesn't provide the `IsInitOnly` property on `IMethodSymbol` so we will need to rely on the LightUp.

**Comment by thomaslevesque** [comment]
> we will need to rely on the LightUp.

I don't know what that means, but if you do, it's all that matters :wink:

**Comment by Evangelink** [comment]
Ahaha sorry talking about implementation details. Basically we cannot follow the latest version of roslyn when it is released because it would force users to always be on the latest version. As we are not on the right version of roslyn we can't see some new elements (in this case that it is a init only property). Because of that we need to rely on some kind of a wrapper which will either return the right roslyn element if it exists or a default value when it doesn't exist.

**Comment by thomaslevesque** [comment]
> Basically we cannot follow the latest version of roslyn when it is released because it would force users to always be on the latest version. As we are not on the right version of roslyn we can't see some new elements (in this case that it is a init only property). Because of that we need to rely on some kind of a wrapper which will either return the right roslyn element if it exists or a default value when it doesn't exist.

Yes, that's what I guessed by looking at the PR ;)
Thanks for the explanation, and for the quick response!

**Comment by twsouthwick** [comment]
I'm still seeing this with the 5.0.200 release of the SDK

**Comment by AmirSasson** [comment]
we are using dotnet version 5.0.401 and still getting false positives on CA2227.
why was this issue closed? @mavasani 
```cs
public record ExtensionSettings(List<Rule> FilterRules);
```
results CA2227
